csrc/scheduler/normalization_utils.cpp

```diff
@@ -677,30 +677,40 @@ getOptionalInnerOuterPersistentBufferBatches(
             scheduler_utils::register_overhead,
         register_per_batch);
   };
-
   const int64_t after_vectorization = inner_dim_numel / vectorize_factor;
   const int64_t threads_per_block_min = std::min(after_vectorization, 128l);
-  const int64_t threads_per_block_max = getThreadsPerSMGivenRegPerThread(255l);
+  const int64_t threads_per_block_max = getThreadsPerSMGivenRegPerThread(
+      scheduler_utils::max_registers_per_thread);
   const int64_t batch_min = getMinimumBatch();
   const int64_t batch_max = getMaximumInnerOuterPersistentBufferBatch();
-
+  // increase batch_max by 1 to allow a small amount of register spills to keep
+  // a small threads_per_block so each SM can host more blocks. This can
+  // increase performance by 10% for edge cases, e.g. hidden size in the range
+  // of (7K,8K]. Disable this optimization if vectorize_factor is 1 due to high
+  // register usage in cases can't be vectorized.
+  const int64_t batch_max_spill_for_occupancy =
+      vectorize_factor > 1 ? batch_max + 1 : batch_max;
   // Start from the smallest threads_per_block. If the corresponding batch size
-  // is larger than batch_max, try increase threads per block by a warp until
-  // the threads_per_block reaches threads_per_block_max or the batch size
-  // reaches batch_min.
+  // is larger than batch_max, try to double threads per block. This will
+  // directly increase threads_per_block from 128 to 256 to avoid intermediate
+  // values such as 160, 192, 224, etc. which can only use a portion of all the
+  // registers on a SM. Because for these intermediate values, one SM can only
+  // host one block if each thread uses 255 registers. Due to the change of
+  // persistent batch size, each threads will not always use 255 registers. But
+  // for the layer norm backward case, we only need to increase threads per
+  // block when persistent batch size is larger than 8. So the smallest batch is
+  // ceilDiv(9,2) = 5, which already uses 216 registers on A100.
```

Comment:
Could you please rephrase this comment? I understand what you wanted to mean, but the comment doesn't seem logically stated. For example:

> So the smallest batch is ceilDiv(9,2) = 5, which already uses 216 registers on A100.

Why "so"? Unclear what this sentence relates to the comment above.